### PR TITLE
feat: Introduce user lookup API [DHIS2-8749]

### PR DESCRIPTION
**Pull request by larshelge**
Introduces a user lookup API. This is intended as a minimal API for user information, which can be used by clients which needs access to basic user information, without exposing sensitive or extensive user information.

The endpoint is located at `/api/userLookup`. The first method located at `/api/userLookup/{id}` allows for lookup by either UID, UUID or username and will return exactly one or zero users.

The second method located at `/api/userLookup?query={query}` allows for querying users by firstname, surname, username and email, and will return many or zero users.

Introduces centralized method for username validation. This needs more work and probably stricter requirements. A first step is to introduce a single place for it to keep it consistent across the system.

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/user/UserServiceTest.java [hunk 1]
```diff
@@ -377,6 +377,28 @@ public void testGetByPhoneNumber()
...
+
+        userService.addUser( userA );
+        userService.addUser( userB );
+        userService.addUser( userC );
+
+        userService.addUserCredentials( credentialsA );
+        userService.addUserCredentials( credentialsB );
+        userService.addUserCredentials( credentialsC );
+
+        //TODO Add tests
```
Comment (netroms) [inline dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/user/UserServiceTest.java:399]: Add missing test?

Comment (larshelge) [reply dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/user/UserServiceTest.java:399]: Yep will do.

File: dhis-2/dhis-support/dhis-support-system/src/main/java/org/hisp/dhis/system/util/ValidationUtils.java [hunk 2]
```diff
@@ -177,6 +179,36 @@ public static boolean urlIsValid( String url )
...
+
+    /**
+     * Validates whether a username is valid.
+     *
+     * @param username the username.
+     * @return true if the username is valid, false otherwise.
+     */
+    public static boolean usernameIsValid( String username )
+    {
+        return username != null && username.length() <= UserCredentials.USERNAME_MAX_LENGTH;
```
Comment (vietnguyen) [inline dhis-2/dhis-support/dhis-support-system/src/main/java/org/hisp/dhis/system/util/ValidationUtils.java:209]: The username string would go directly into the query `from UserCredentials uc where uc.username = :username`. Have we already had sql escape somewhere in the api ? 
e.g. what if username string =  (select username from users order by lastlogin desc limit 1)

Comment (larshelge) [reply dhis-2/dhis-support/dhis-support-system/src/main/java/org/hisp/dhis/system/util/ValidationUtils.java:209]: Good point. We retrieve user credentials in `HibernateUserCredentialsStore` using a Hibernate `Query` with named parameters, where SQL is automatically sanitized.

**Pull request merged**